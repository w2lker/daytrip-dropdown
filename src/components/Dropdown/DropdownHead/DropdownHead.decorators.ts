import withStyles from "react-jss";

import DropdownHead from "./DropdownHead";
import dropdownHeadStyles from "./DropdownHead.styles";

const styledComponent = withStyles(dropdownHeadStyles)(DropdownHead);

export default styledComponent;
