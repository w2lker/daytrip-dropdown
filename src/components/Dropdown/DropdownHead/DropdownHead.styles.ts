const dropdownHeadStyles = {
  wrapper: {
    background: '#ECF0F2',
    boxShadow: '0 2px 9px 0 rgba(146,166,176,0.15)',
    borderRadius: 4,
    position: 'relative',
    padding: '16px 30px'
  },
  label: {
    paddingBottom: 6,
    fontWeight: 500,
    fontSize: 10,
    color: '#445A64'
  },
  content: {
    fontSize: 16,
    fontWeight: 400,
    color: '#7A929E',
    letterSpacing: 0,
  },
  caret: {
    position: 'absolute',
    right: 15,
    top: '50%',
    transform: 'translateY(-50%)',
    transition: 'transform 0.3s ease-in-out'
  },
  caretReverted: {
    transform: 'translateY(-50%) rotate(-180deg)'
  }
};

export default dropdownHeadStyles;
