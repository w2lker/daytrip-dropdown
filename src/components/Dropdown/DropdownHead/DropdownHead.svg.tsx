import React, {ReactElement} from "react";

export const chevron = (color: string = '#617E8D'): ReactElement => {
  return (
    <svg width="12px" height="8px" viewBox="0 0 12 8">
      <polygon fill={color} points="1.41 0 6 4.59 10.59 0 12 1.42 6 7.42 0 1.42"/>
    </svg>
  );
};
