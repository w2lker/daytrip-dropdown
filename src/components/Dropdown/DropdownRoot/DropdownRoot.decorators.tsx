import withStyles from "react-jss";

import DropdownRoot from "./DropdownRoot";
import DropdownLang from "./DropdownRoot.lang";

import withControlledSwitcher from "../../Decorators/withControlledSwitcher";
import withLangContext from "../../Decorators/withLangContext";

import dropdownRootStyles from "./DropdownRoot.styles";
import lang from "../../../const/lang";

const styledComponent = withStyles(dropdownRootStyles)(DropdownRoot);
const withControlledSwitcherDecorated = withControlledSwitcher(styledComponent);
const withLangContextDecorated = withLangContext(withControlledSwitcherDecorated, DropdownLang, lang.dropdown);

export default withLangContextDecorated;
