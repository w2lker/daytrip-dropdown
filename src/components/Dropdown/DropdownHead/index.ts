import DropdownHead from "./DropdownHead.decorators";

export default DropdownHead;
