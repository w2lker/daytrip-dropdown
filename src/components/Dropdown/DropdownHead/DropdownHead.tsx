import React, { useContext } from "react";

import {WithStyles} from "react-jss";
import classNames from "classnames";

import DropdownLang from "../DropdownRoot/DropdownRoot.lang";

import dropdownHeadStyles from "./DropdownHead.styles";
import lang from "../../../const/lang";

import { IDropdownOptionsElement } from "../DropdownRoot/DropdownRoot";
import {getValue} from "../../../utils/dropdown";
import {chevron} from "./DropdownHead.svg";

export interface IDropdownHeadProps extends WithStyles<typeof dropdownHeadStyles>{
  opened: boolean;
  label?: string;
  placeholder?: string;
  selectedItem?: IDropdownOptionsElement | null,
  onClick: () => void;
}

const DropdownHead: React.FC<IDropdownHeadProps> = (props) => {
  const { opened, label, placeholder, selectedItem, onClick, classes } = props;
  // Fallback if there is no context provided
  const contextLang = useContext(DropdownLang) as typeof lang.dropdown;
  const defaultTexts = contextLang && contextLang.head ? contextLang.head : { label: '', placeholder: '' };

  const labelTest = label || defaultTexts.label;
  const contentPlaceholder = placeholder || defaultTexts.placeholder;
  const contentText = selectedItem ? getValue(selectedItem) : contentPlaceholder;
  const caretClassname = classNames({
    [classes.caret]: true,
    [classes.caretReverted]: opened,
  });
  const handleMouseDown = (event: React.MouseEvent): void => {
    if (!event.button) {
      return onClick();
    }
  };

  return (
    <div
      className={classes.wrapper}
      onMouseDown={handleMouseDown}
    >
      <div className={classes.label}>{labelTest}</div>
      <div className={classes.content}>{contentText}</div>
      <div className={caretClassname}>
        {chevron()}
      </div>
    </div>
  )
};

export default DropdownHead;
