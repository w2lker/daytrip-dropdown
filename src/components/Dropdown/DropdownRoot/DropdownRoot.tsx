import React, {createContext, useEffect, useState} from 'react';

import DropdownHead from "../DropdownHead";
import DropdownBody from "../DropdownBody";

import {modifyDropdownOptionsDuplicates, modifyDropdownOptionsMultiline} from './DropdownRoot.helpers';
import {getKey} from "../../../utils/dropdown";
import lang from "../../../const/lang";

export interface IDropdownOptionsElement {
    [value: string]: string;
}

export type IDropdownOptionsArray = IDropdownOptionsElement[];

interface IDropdownProps {
    options: IDropdownOptionsArray;
    selected?: string;
    rows?: number;
    label?: string;
    placeholder?: string;
    onSelect?: (selected: string) => void;
    // TODO: define class definition in JSS package
    classes?: any;
}

export const DropdownLang = createContext({});

const DropdownRoot: React.FC<IDropdownProps> = (props) => {
    const { options, rows, label, placeholder, selected, classes, onSelect } = props;
    let modifiedOptions: IDropdownOptionsArray;
    // Filter options array to display only unique key value to prevent issue with onSelect callback validation
    useEffect(() => {
        const withoutDuplicates = modifyDropdownOptionsDuplicates(options);
        modifiedOptions = modifyDropdownOptionsMultiline(withoutDuplicates);
    }, [options]);

    const [isOpened, setIsOpened] = useState(false);

    const selectedOption = selected && options && options.length ?
        options.find( (opt) => getKey(opt) === selected)
    : null;

    return (
      <DropdownLang.Provider value={lang.dropdown}>
        <div className={classes.root}>
            <DropdownHead
                opened={isOpened}
                label={label}
                placeholder={placeholder}
                onClick={() => setIsOpened(!isOpened)}
                selectedItem={selectedOption}
            />
            {isOpened && (
                <DropdownBody
                  selected={selected}
                  options={options}
                  rows={rows}
                  onSelect={onSelect}
                  onClose={() => setIsOpened(false)}
                />
            )}
        </div>
      </DropdownLang.Provider>
    );
};

DropdownRoot.defaultProps = {
    options: [],
    rows: 5,
    classes: {
        root: '',
    }
};

export default DropdownRoot;
