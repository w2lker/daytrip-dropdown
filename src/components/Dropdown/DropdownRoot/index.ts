import Dropdown from "./DropdownRoot.decorators";

export default Dropdown;
