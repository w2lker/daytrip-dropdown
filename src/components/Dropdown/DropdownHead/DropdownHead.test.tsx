import React from 'react';

import { mount, shallow } from 'enzyme';

import DropdownHead, { IDropdownHeadProps } from './DropdownHead';
import DropdownHeadDecorated from './DropdownHead.decorators';
import dropdownHeadStyles from './DropdownHead.styles';
import {getValue} from "../../../utils/dropdown";
import {DropdownLang, IDropdownOptionsElement} from "../DropdownRoot/DropdownRoot";

const setup = () => {
  const classes = {
    wrapper: 'dropdownHead-wrapper',
    label: 'dropdownHead-label',
    content: 'dropdownHead-content',
    caret: 'dropdownHead-caret',
    caretReverted: 'dropdownHead-caretReverted',
  };
  const classesKeys = Object.keys(classes);
  const sampleProps: IDropdownHeadProps = {
    opened: true,
    label: 'Test label',
    placeholder: 'Test Placeholder',
    selectedItem: {'key1': 'value1'},
    onClick: jest.fn(),
    classes,
  };

  const defaultText = {
    head: {
      label: 'default label',
      placeholder: 'default placeholder'
    }
  };

  return { sampleProps, classes, classesKeys, defaultText };
};

describe('DropdownHead component', () => {

  it('match snapshot', () => {
    const { sampleProps } = setup();
    const component = shallow(<DropdownHead {...sampleProps} />);
    expect(component.debug()).toMatchSnapshot();
  });

  it('displays content', () => {
    const { classes, sampleProps } = setup();
    const component = shallow(<DropdownHead {...sampleProps} />);
    expect(component.find(`.${classes.label}`).text()).toBe(sampleProps.label);
    expect(component.find(`.${classes.content}`).text()).toBe(getValue(sampleProps.selectedItem as IDropdownOptionsElement));
    if (sampleProps.opened) {
      expect(component.find(`.${classes.caretReverted}`).length).toBe(1);
    }
  });

  it('displays placeholder', () => {
    const { classes, sampleProps } = setup();
    const props = {
      ...sampleProps,
      selectedItem: null,
    };
    const component = shallow(<DropdownHead {...props} />);
    expect(component.find(`.${classes.label}`).text()).toBe(sampleProps.label);
    expect(component.find(`.${classes.content}`).text()).toBe(sampleProps.placeholder);
  });

  it('displays constant text', () => {
    const { classes, sampleProps, defaultText } = setup();
    const props = {
      ...sampleProps,
      selectedItem: undefined,
      label: undefined,
      placeholder: undefined,
    };
    const component = mount(
      <DropdownLang.Provider value={defaultText}>
        <DropdownHead {...props} />
      </DropdownLang.Provider>
    );

    expect(component.find(`.${classes.label}`).text()).toBe(defaultText.head.label);
    expect(component.find(`.${classes.content}`).text()).toBe(defaultText.head.placeholder);
  });

  it('handles click', () => {
    const { classes, sampleProps } = setup();
    const component = shallow(<DropdownHead {...sampleProps}/>);
    component.find(`.${classes.wrapper}`).simulate('click');
    expect(sampleProps.onClick).toBeCalledTimes(1);
  });
});

describe('DropdownHead styles', () => {
  it('contains principle classes', () => {
    const { classesKeys } = setup();
    expect(dropdownHeadStyles).toBeDefined();
    classesKeys.forEach((keyValue) => {
      // @ts-ignore
      expect(dropdownHeadStyles[keyValue]).toBeDefined();
    });
  });
});

describe('DropdownHead decorators', () => {
  it('provides styled classes from decorators', () => {
    const { sampleProps, classesKeys } = setup();
    const component = mount(<DropdownHeadDecorated {...sampleProps} />);
    // @ts-ignore
    const assignedClasses = component.find('NoRenderer').props().classes;
    classesKeys.forEach(keyValue => {
      expect(assignedClasses[keyValue]).toBeDefined();
    });
  });
});
