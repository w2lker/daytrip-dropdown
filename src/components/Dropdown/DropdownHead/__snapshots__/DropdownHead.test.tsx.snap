// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownHead component match snapshot 1`] = `
"<div className=\\"dropdownHead-wrapper\\" onMouseDown={[Function: handleMouseDown]}>
  <div className=\\"dropdownHead-label\\">
    Test label
  </div>
  <div className=\\"dropdownHead-content\\">
    value1
  </div>
  <div className=\\"dropdownHead-caret dropdownHead-caretReverted\\">
    <svg width=\\"12px\\" height=\\"8px\\" viewBox=\\"0 0 12 8\\">
      <polygon fill=\\"#617E8D\\" points=\\"1.41 0 6 4.59 10.59 0 12 1.42 6 7.42 0 1.42\\" />
    </svg>
  </div>
</div>"
`;
